### PR TITLE
Fix Docker Compose & Add Boilerplate Project

### PR DESCRIPTION
**Pull request by HassanAbouelela**
This PR works on the developer experience for bot-core by adding better infrastructure and boilerplate to help developers test out their changes. This is achieved by adding prominent documentation which explains how to test features in sample and real projects, adding a sample project which contains all the boilerplate for rapid prototyping, and a fixed docker configuration which should work better for this project.

I also added aiodns as an explicit dependency. Our downstream projects already have it so they don't encounter issues, but for local testing it had to be added. We should remove it downstream as well.

### CODE REVIEW

**Comment by ChrisLovering** [comment]
@HassanAbouelela rebase

File: dev/README.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+The steps below should provide most of the groundwork you need, but the exact requirements will
+vary by the feature you're working on.
+
+
+Option 1
+--------
+1. Navigate to the project you want to install bot-core in, such as bot or sir-lancebot
+2. Run ``pip install /path/to/botcore`` in the project's environment
+
+   - The path provided to install should be the root directory of this project on your maching.
```
Comment (ChrisLovering) [inline dev/README.rst:22]:
```suggestion
   - The path provided to install should be the root directory of this project on your machine.
```

File: dev/README.rst [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+   - Make sure to install in the correct environment. Most Python Discord projects use
+     poetry, so you can run ``poetry run pip install /path/to/botcore``.
+
+3. You can now use features from your local bot-core changes.
+   To load new changes, run the install command again.
+
+
+Option 2
+--------
+1. Copy the :repo-file:`bot template folder <dev/bot>` to the root of your project.
```
Comment (ChrisLovering) [inline dev/README.rst:33]:
```suggestion
1. Copy the :repo-file:`bot template folder <dev/bot>` to the root of the bot-core project.
```

File: dev/bot/__main__.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+from . import Bot
+
+dotenv.load_dotenv()
+botcore.utils.apply_monkey_patches()
+
+roles = os.getenv("ALLOWED_ROLES")
+roles = [int(role) for role in roles.split(",")] if roles else []
+
+bot = Bot(
+    guild_id=int(os.getenv("GUILD_ID")),
```
Comment (ChrisLovering) [inline dev/bot/__main__.py:19]: Should we document these required env vars? Same goes for "TOKEN" below.

ALLOWED_ROLES & PREFIX could be documented too, even though optional.

**Review by ChrisLovering — CHANGES_REQUESTED** [review]

File: dev/bot/__main__.py (hunk 3 above)
Comment (HassanAbouelela) [reply dev/bot/__main__.py:19]: Yeah that's a good idea

Comment (HassanAbouelela) [reply dev/bot/__main__.py:19]: 69ae0b0b

**Review by ChrisLovering — APPROVED** [review]

**Pull request merged**